portal/portal-charon/charon/src/webapp/scripts/themeswitcher.js

```diff
@@ -1,43 +1,97 @@
+    
+    function init() {
+        // attach listener to dark theme toggle switch
+        darkThemeSwitcher && darkThemeSwitcher.addEventListener('click', toggleDarkTheme, false);
+
+        if(isLoggedIn()) {
+            // only check for unset theme preference because light and dark themes are already set by Java
+            if(isPortalThemeUserPrefUnset()) {
+                if(isOsDarkThemeSet()) {
+                    enableDarkTheme();
+                } else {
+                    // to define a user preference:
+                    setPortalThemeUserPref(lightThemeClass);
+                }
+            }
+        } else if(isOsDarkThemeSet()) {
+            addCssClassToMarkup(darkThemeClass);
+        }
         
-        initTheme(); // on page load, if user has already selected a specific theme -> apply it
+        if (document.documentElement.classList.contains(darkThemeClass)) {
+            // the dark theme switch toggle is off by default, so toggle to on if dark theme is enabled
+            setDarkThemeSwitcherToggle(true);
+        }
+    }
 
-        themeSwitcher.addEventListener('click', setTheme, false);
+    function addCssClassToMarkup(themeClass) {
+        document.documentElement.classList.add(themeClass);
+    }
 
+    function removeCssClassFromMarkup(themeClass) {
+        document.documentElement.classList.remove(themeClass);
     }
 
-    function setTheme() {
+    function isOsDarkThemeSet() {
+        return window.matchMedia('(prefers-color-scheme: dark)').matches;
+    }
+
+    function isLoggedIn() {
+        return portal.user.id;
+    }
 
-        // Set to dark mode
-        if (themeSwitcher.getAttribute("aria-checked") === "true") {
-            themeSwitcher.setAttribute("aria-checked", "false");
-            document.firstElementChild.classList.add('sakai-dark-theme');
-            localStorage.setItem('sakai-theme', 'dark');
+    function isPortalThemeUserPrefUnset() {
+        if (portal.userTheme === "") {
+            return true;
+        } else {
+            return false;
         }
-        // Set to light mode
-        else {
-            themeSwitcher.setAttribute("aria-checked", "true");
-            document.firstElementChild.classList.remove('sakai-dark-theme');
-            localStorage.setItem('sakai-theme', 'light')
+    }
+
+    function setDarkThemeSwitcherToggle(onOff) {
+        darkThemeSwitcher.setAttribute("aria-checked", onOff);
+    }
+
+    function setPortalThemeUserPref(theme) {
+        var url = '/direct/userPrefs/updateKey/' + portal.user.id + '/sakai:portal:theme?theme=' + theme;
+        var ajaxRequest = new XMLHttpRequest();
+        ajaxRequest.open("PUT", url);
+        ajaxRequest.send();
+    }
+
+    function toggleDarkTheme() {
+        // Set to dark theme
+        if (darkThemeSwitcher.getAttribute("aria-checked") === "false") {
+            enableDarkTheme();
+        } else {
+            // Set to light theme
+            enableLightTheme();
```

Comment:
darkThemeSwitcher.getAttribute("aria-checked") === "false" ? enableDarkTheme() : enableLightTheme();

Reply:
Thanks for the suggestion. I've updated this line now.